### PR TITLE
Add expectation value function to simulator

### PR DESCRIPTION
**Pull request by thomashaener**
Enables the simulator to calculate expectation values w.r.t. `QubitOperator`.

### CODE REVIEW

File: projectq/backends/_sim/_cppkernels/simulator.hpp [hunk 1]
```diff
@@ -232,6 +242,38 @@ class Simulator{
...
+        Fusion::Matrix Z = {{1., 0.}, {0., -1.}};
+        std::vector<Fusion::Matrix> gates = {X, Y, Z};
+
+        calc_type expectation = 0.;
+        auto current_state = vec_;
+        for (auto const& term : td){
+            auto const& coefficient = term.second;
+            for (auto const& local_op : term.first){
+                unsigned id = ids[local_op.first];
+                apply_controlled_gate(gates[local_op.second - 'X'], {id}, {});
```
Comment (damiansteiger) [inline projectq/backends/_sim/_cppkernels/simulator.hpp]: :laughing: 
well done Sir 💯

File: projectq/backends/_sim/_pysim.py [hunk 2]
```diff
@@ -228,6 +228,33 @@ def emulate_math(self, f, qubit_ids, ctrlqubit_ids):
...
+        Y = [[0., -1j], [1j, 0.]]
+        Z = [[1., 0.], [0., -1.]]
+        gates = [X, Y, Z]
+        expectation = 0.
+        current_state = _np.copy(self._state)
+        for (term, coefficient) in terms_dict:
+            for local_op in term:
+                qb_id = ids[local_op[0]]
+                self.apply_controlled_gate(gates[ord(local_op[1]) - ord('X')],
+                                           [qb_id], [])
```
Comment (damiansteiger) [inline projectq/backends/_sim/_pysim.py]: Not a single `if` too much 👍

File: projectq/backends/_sim/_simulator_test.py [hunk 3]
```diff
@@ -191,6 +195,47 @@ def test_simulator_emulation(sim):
...
+    op2 = QubitOperator('Y0')
+    expectation = sim.get_expectation_value(op2, qureg)
+    assert 1. == pytest.approx(expectation)
+    Z | qureg[0]
+    expectation = sim.get_expectation_value(op2, qureg)
+    assert -1. == pytest.approx(expectation)
+
+    op_sum = QubitOperator('Y0 X1 Z2') + QubitOperator('X1')
+    op_sum += op_sum
+    op_sum /= 2.
```
Comment (damiansteiger) [inline projectq/backends/_sim/_simulator_test.py]: What is the purpose of this?
`(op_sum + op_sum) / 2. == op_sum`

Comment (thomashaener) [inline projectq/backends/_sim/_simulator_test.py]: I will remove that, since operations are already tested in qubit_operator.

**Pull request merged**